src/algebra/cubic_solution.lean

```diff
@@ -0,0 +1,570 @@
+/-
+Copyright (c) 2021 Nicholas Dyson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Nicholas Dyson.
+-/
+
+import algebra.invertible
+import tactic.linarith
+import tactic.field_simp
+import order.filter.at_top_bot
+
+/-!
+# Cubic solutions
+
+In this file we define and prove the three solutions to a general cubic in a general
+field that is not of characteristic two or three, and uniqueness of these solutions.
+
+## Main results
+
+`cubic_solution` : The solution to a cubic
+`cubic_solution_unique` : The three solutions generated by using the three cube roots in the
+                          formula are the only solutions.
+-/
+
+
+section fields
+variables {F : Type*}
+
+/--
+((-1 + sqrt(-3))/2)^3 = 1
+-/
+def cube_root_unity_a [field F] (root_three_i : F) : F :=
+  (-1 + root_three_i)/2
+
+/--
+((-1 - sqrt(-3))/2)^3 = 1
+-/
+def cube_root_unity_b [field F] (root_three_i : F) : F :=
+  (-1 - root_three_i)/2
+
+lemma multiply_out_cubed [field F] (x y : F) : (x + y) * (x + y) * (x + y) =
+        (x * x * x) + (3 * x * x * y) + (3 * x * y * y) + (y * y * y) :=
+begin
+  ring,
+end
+
+/--
+If F isn't of characteristic 2, 2^3 ≠ 0 also.
+--/
+instance four_ne_zero [field F] [invertible (2 : F)] : invertible (4 : F) :=
+begin
+    have cubed : (2 : F) ^ 2 = (4 : F),
+      ring,
+    rw <- cubed, exact invertible_pow 2 2,
+end
+
+/--
+If F isn't of characteristic 2, 2^3 ≠ 0 also.
+--/
+instance eight_ne_zero [field F] [invertible (2 : F)] : invertible (8 : F) :=
+begin
+    have cubed : (2 : F) ^ 3 = (8 : F),
+      ring,
+    rw <- cubed, exact invertible_pow 2 3,
+end
+
+/--
+If F isn't of characteristic 3, 3^3 ≠ 0 also.
+--/
+instance twenty_seven_ne_zero [field F] [invertible (3 : F)] : invertible (27 : F) :=
+begin
+    have cubed : (3 : F) ^ 3 = (27 : F),
+      ring,
+    rw <- cubed, exact invertible_pow 3 3,
+end
+
+/--
+((-1 + sqrt(-3))/2)^3 = 1
+--/
+@[simp] lemma cubrt_unity_a_correct [field F] [invertible (2 : F)] (root_three_i : F)
+    (rt3_i_correct : root_three_i * root_three_i = (-3))  :
+    (cube_root_unity_a root_three_i) * (cube_root_unity_a root_three_i) * (cube_root_unity_a root_three_i)
+     = 1 :=
+begin
+  unfold cube_root_unity_a,
+  have simp_numerator : (-1 + root_three_i) * (-1 + root_three_i) * (-1 + root_three_i) = 8,
+    rw multiply_out_cubed, rw rt3_i_correct,
+    repeat {rw mul_neg_one}, repeat {rw neg_neg},
+    rw mul_assoc (-3) root_three_i root_three_i, rw rt3_i_correct,
+    ring,
```

Comment:
Mathlib style is to use `{}` for subgoals, because this ensures that your indents aren't lying about where the subgoal ends.
```suggestion
  { rw multiply_out_cubed, rw rt3_i_correct,
    repeat {rw mul_neg_one}, repeat {rw neg_neg},
    rw mul_assoc (-3) root_three_i root_three_i, rw rt3_i_correct,
    ring, },
```